### ISSUE TITLE
ohdsi - Postgresql - Issues deploying to K8s

### ISSUE THREAD

**Issue by ssaranathan** [chgl/charts#34]
Hi @chgl ,
As you mentioned, I am moving this topic from https://github.com/OHDSI/Broadsea/issues/23

Two main issues that I saw in the Postgresql deployment:

1.  No cpu limit specified in the values: and hence I got this message: 
create Pod ohdsi-postgresql-0 in StatefulSet ohdsi-postgresql failed error: Pod "ohdsi-postgresql-0" is invalid: spec.containers[0].resources.requests: Invalid value: "250m": must be less than or equal to cpu limit

So I am trying to update the values to include this.

2.  mkdir: cannot create directory ‘/bitnami/postgresql/data’: Permission denied
I am looking into this issue.  Maybe I should create a PVC first and use that in the values.

I will update once I do these.

Thanks,

**Comment by chgl** [comment]
Thank you for creating the issue!

The first problem looks like some sort of admission controller enforcing resource limits for all pods.

I am using Bitnami's Postgres as a sub-chart for OHDSI. You can customize any value of this chart under the `postgresql`-key, eg:

```yaml
postgresql:
  # see https://github.com/bitnami/charts/blob/master/bitnami/postgresql/values.yaml#L626
  resources:
    requests:
      memory: 256Mi
      cpu: 250m
    limits:
      memory: 256Mi
      cpu: 250m
```
These values are documented over at: https://github.com/bitnami/charts/blob/master/bitnami/postgresql/values.yaml

> 2.

Permission errors for volumes are always annoying to resolve. I am running an instance on-premise using VMWare's `vsphere-storage`, the following helped in getting it to work even when running as non-root:

```yaml
  postgresql:
    enabled: true
    persistence:
      size: 128Gi
      storageClass: "vsphere-storage"
    volumePermissions: # this is important
      enabled: true 
      securityContext: 
        runAsUser: 1001
        runAsGroup: 1001
    shmVolume: # this as well
      chmod:
        enabled: false
```

See also https://github.com/bitnami/charts/issues/1210#issuecomment-688128526 if it doesnt work.

**Comment by ssaranathan** [comment]
Hi @chgl , 
Thanks for your tips.  I tried the following and still got errors:
====================
 postgresql:
 
  enabled: true
  image: # +doc-gen:break
    tag: 13.1.0
    pullPolicy: Always
  postgresqlDatabase: "ohdsi"
  existingSecret: "pg-atlas-secret"
  resources:
    requests:
      memory: 256Mi
      cpu: 250m
    limits:
      cpu: 250m
  persistence:
    enabled: true
    existingClaim: ohdsi-pg-claim
    mountPath: /bitnami/postgresql/data
    volumePermissions: # this is important
      enabled: true 
      securityContext: 
        runAsUser: 1001
        runAsGroup: 1001
    shmVolume: # this as well
      chmod:
        enabled: false
  replication:
    enabled: false
    readReplicas: 1
    synchronousCommit: "on"
    numSynchronousReplicas: 1
  metrics:
    enabled: false
  containerSecurityContext:
    allowPrivilegeEscalation: false
    capabilities:
      drop:
        - ALL

===================
Error:

postgresql 23:38:09.47 
postgresql 23:38:09.49 Welcome to the Bitnami postgresql container
postgresql 23:38:09.49 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql
postgresql 23:38:09.51 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql/issues
postgresql 23:38:09.51 
postgresql 23:38:09.63 INFO  ==> ** Starting PostgreSQL setup **
postgresql 23:38:09.73 INFO  ==> Validating settings in POSTGRESQL_* env vars..
postgresql 23:38:09.79 INFO  ==> Loading custom pre-init scripts...
postgresql 23:38:09.83 INFO  ==> Initializing PostgreSQL database...
chmod: changing permissions of '/bitnami/postgresql/data': Operation not permitted
postgresql 23:38:09.97 WARN  ==> Lack of permissions on data directory!
chmod: changing permissions of '/bitnami/postgresql/data': Operation not permitted
postgresql 23:38:09.97 WARN  ==> Lack of permissions on data directory!
postgresql 23:38:09.97 INFO  ==> pg_hba.conf file not detected. Generating it...
postgresql 23:38:10.01 INFO  ==> Generating local authentication configuration

**Comment by chgl** [comment]
Might just be a copy-and-paste error, but `volumePermissions` shouldn't be nested inside the `persistence`-key but should be just below `postgresql`.

If that didn't fix it, you can try this: https://github.com/helm/charts/issues/20923#issuecomment-733078301

**Comment by ssaranathan** [comment]
Thanks for catching that @chgl .  Little progress now.  I now see this in the pod log:

`container "ohdsi-postgresql" in pod "ohdsi-postgresql-0" is waiting to start: PodInitializing`

I tried the fix  kubectl exec -it ohdsi-postgresql-0 -- chmod 700 /bitnami/postgresql -R -n ohdsi
Error from server (NotFound): pods "ohdsi-postgresql-0" not found

Since the Pod is not up, it gives the above error

**Comment by chgl** [comment]
Can you check the log output of the init-chmod-data container? It might be looping or crashing for some reason thus preventing the main postgres container from starting. It could also be that the existing volume claim (it's underlying volume to be precise) might have some incompatible permissions at this point. If possible, please also try creating a new volume claim of a different name.

**Comment by ssaranathan** [comment]
Hi @chgl ,
Thanks for your responses.  Yes, every time, before I install the chart, I delete the PVC and provision it again.  

This is what I see in the init-container log:

```
+ chown 1001:1001 /bitnami/postgresql/data
chown: changing ownership of '/bitnami/postgresql/data': Operation not permitted
+ mkdir -p /bitnami/postgresql/data/data
mkdir: cannot create directory '/bitnami/postgresql/data/data': Permission denied
+ chmod 700 /bitnami/postgresql/data/data
chmod: cannot access '/bitnami/postgresql/data/data': No such file or directory
+ find /bitnami/postgresql/data -mindepth 1 -maxdepth 1 -not -name conf -not -name .snapshot -not -name lost+found
+ xargs chown -R 1001:1001
chown: missing operand after '1001:1001'
Try 'chown --help' for more information.
```

**Comment by chgl** [comment]
That's frustrating. Unfortunately since permission handling is so dependent on the underlying Volume provisioner, I can only offer trial-and-error solutions. So here's the next attempt: try setting

```yaml
postgresql:
  volumePermissions:
     securityContext: 
       runAsUser: 0 # this should run the init container as root which should have ample permissions.
```

What Kuebrnetes distribution and storage plugins are you using?

Note also the comment in the postgres chart https://github.com/bitnami/charts/blob/master/bitnami/postgresql/values.yaml#L66 :

```yaml
  ## Init container Security Context
  ## Note: the chown of the data folder is done to securityContext.runAsUser
  ## and not the below volumePermissions.securityContext.runAsUser
  ## When runAsUser is set to special value "auto", init container will try to chwon the
  ## data folder to autodetermined user&group, using commands: `id -u`:`id -G | cut -d" " -f2`
  ## "auto" is especially useful for OpenShift which has scc with dynamic userids (and 0 is not allowed).
  ## You may want to use this volumePermissions.securityContext.runAsUser="auto" in combination with
  ## pod securityContext.enabled=false and shmVolume.chmod.enabled=false
  ##
  securityContext:
    runAsUser: 0
```

>  ## You may want to use this volumePermissions.securityContext.runAsUser="auto" in combination with
>  ## pod securityContext.enabled=false and shmVolume.chmod.enabled=false

Would be my next suggestion.

**Comment by ssaranathan** [comment]
Thanks @chgl.  I tried the first option and Postgres came up ..  
I will now look into the other components of the stack.  
Thanks a lot and appreciate your prompt help. 


```

postgresql:
  volumePermissions:
     securityContext: 
       runAsUser: 0 # this should run the init container as root which should have ample permissions.
```

**Comment by chgl** [comment]
@ssaranathan glad to hear 🙂. Let me know how it goes!

**Comment by ssaranathan** [comment]
Hi @chgl ,
Sorry for the delay in responding back.

After fixing the Postgresql root issue, the Postgres pod came up fine.

However, the WebAPI pod is continuously crashing and I see this error message in the deployment:
```
 Startup probe failed: Get "http://172.30.225.219:8080/WebAPI/info": dial tcp 172.30.225.219:8080: connect: connection refused
Back-off restarting failed container
```

Atlas is showing Green.. however, I am unable to access the application.  I am trying to see if it is the Ingress that is causing the issue.

**Comment by chgl** [comment]
Hm, is there any information in the WebAPI logs? It could be that starting the WebAPI (or more specifically applying the WebAPI migrations) takes a bit longer than I had anticipated when setting the startupProbe delay. You can try increasing it to say 300s (5 minutes):

https://github.com/chgl/charts/blob/master/charts/ohdsi/values.yaml#L142

```yaml
webApi:
  startupProbe:
    initialDelaySeconds: 300
    periodSeconds: 60 # you can increase this as well if the WebAPI takes such a long time to finish setting up
```

**Comment by ssaranathan** [comment]
Hi @chgl ,
In order to move forward, I have temporarily disabled Postgres deployment.  I have instead connected to a Postgres instance in the cloud and provided that information in the db section of WebAPI.

I have also increased the probe timings to large values to they don't even get triggered until I fix the remaining issues.

After doing those, I still see WebAPI crashing.  One thing that I noticed is, I don't see the following values passed to WebAPI:

I used to have these in my WebAPI when I am not installing via the helm chart.  

Don't we need this in the helm chart ?

```
datasource.cdm.schema: 
  datasource.ohdsi.schema: 
  spring.jpa.properties.hibernate.default_schema: 
  flyway_schemas: ohdsi
  flyway.placeholders.ohdsiSchema: 
```

**Comment by chgl** [comment]
The integration tests and my local installation seem to work without them - but in case your existing postgres DB uses different naming it might indeed be an issue. You can try setting them using the `webApi.extraEnv:` key, eg:

```yaml
wepApi:
  extraEnv:
    - name: DATASOURCE_CDM_SCHEMA
      value: cdm
    # ...
```

The logs in https://github.com/chgl/charts/runs/1895133739?check_suite_focus=true#step:8:2017 show a successfully initialized WebAPI instance as part of the ci runs. I definitely can't rule out that it only works because of some default values I assume to be set. So the additional env vars you mentioned could be a clue and an oversight by me!

**Comment by ssaranathan** [comment]
Hi @chgl ,
After adding the extraEnv variables, I can see WebAPI pod coming up well.  The pod doesn't crash anymore.   I am still connecting to a PostgresDB that I have created on the cloud.
```
2021-02-27 16:36:15.237 INFO main org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedServletContainer -  - Tomcat started on port(s): 8080 (http)
2021-02-27 16:36:15.253 INFO main org.ohdsi.webapi.WebApi -  - Started WebApi in 44.392 seconds (JVM running for 45.786)
```

Now, when I try to access the WebAPI (https:/xxx/WebAPI/info), I don't see any response and the browser is just spinning.  

Since I cannot access WebAPI, Atlas still shows the message that it is unable to initialize.

Ingress is setup in the cluster.

Anything else that I could have missed ?  Thank you very much for all your responses and help.

**Comment by chgl** [comment]
I'm grateful for your patience and help as well! I've just added these env vars via https://github.com/chgl/charts/pull/37 available as chart version 0.4.7 - they are all set to the same value from `webApi.db.schema`, which defaults to `ohdsi`.

Can you try accessing the WebApi/info via curl to see what exactly it returns?:

```sh
curl -vvv https:/xxx/WebAPI/info
```

Also try accessing the WebAPI using `kubectl port-forward` (https://kubernetes.io/docs/tasks/access-application-cluster/port-forward-access-application-cluster/) to see if the response looks healthy. What Ingress controller are you using?

**Comment by ssaranathan** [comment]
Hi @chgl ,
Thanks for your prompt responses.  We use F5 Ingress controller.

curl -v xxx/WebAPI/info did not give any output.  It is just stuck.

Also, I was just curious to get the status of the container..   The POD comes up fine and I can see the logs as shown in my previous comments.  But the container is in NOT Ready status.. :-(

```
NAME                            READY   STATUS    RESTARTS   AGE
ohdsi-atlas-fcb7fc5cb-9c6cb     1/1     Running   0          21m
ohdsi-webapi-7685b7f5d7-qvm9h   0/1     Running   0          21m
```
![image](https://user-images.githubusercontent.com/66810237/109403016-afcc8100-7928-11eb-9133-6f708eae42fd.png)

**Comment by ssaranathan** [comment]
Hi @chgl ,
One more thing..
In the webapi -> db section of the values.yml, I specified the username and password and I left the secret and secret key empty strings.  However, when I do a pod describe, I see this:

`      DATASOURCE_PASSWORD:                             <set to the key 'postgresql-password' in secret 'ohdsi-webapi-db-secret'>  Optional: false`

Is this an issue ?

**Comment by ssaranathan** [comment]
Hi @chgl ,
Made some good progress.  I think the WebAPI container was in not ready state because I had set the readiness probe times to a big value.  I reduced it and once it ran, I see the container in the ready state and I was able to invoke the WebAPI/info and other apis.  I also had to delete the ingress and reinstall it after WebAPI came up.  Without that, the WebAPI was not responding.

Now, I can curl to both /atlas as well as WebAPI/info.

The issue that I see in the Atals UI is related to CORS 
`
 Access to fetch at 'https://xxx/WebAPI/source/sources' from origin 'http://xxx' has been blocked by CORS policy: The 'Access-Control-Allow-Origin' header has a value 'http://localhost' that is not equal to the supplied origin. Have the server send the header with a valid value, or, if an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.`

I did check the /usr/share/nginx/html/atlas/js/config-local.js and it correctly shows the WebAPI end point.  Not sure where localhost is coming up.

**Comment by ssaranathan** [comment]
Hi @chgl ,
I was able to resolve the above issue by adding an extraEnv variable : WEBAPI_URL in the atlas section of the values.yml.  You might want to add that in your instructions.

Now the only other thing that is pending is the issue with Postgres creation using the bitnami/Postgres chart.  After the postgres pod comes up as root user, when I go that pod and run psql -U , it does not authenticate the user that I have defined in the secret.

**Comment by chgl** [comment]
You can already override this env var using the `atlas.webApiUrl` value.

If the ingress is enabled, then `WEBAPI_URL` of the atlas container is set to `ingress.hosts[0].host`/WebAPI by default. Which I was hoping would work in most cases. To what value was the WEBAPI_URL set before you manually overrode it (it should actually still be set, if you inspect the pod's definition)? 

I've now added values to configure CORS for the WebAPI - but it apparently wasn't this issue.

**Comment by ssaranathan** [comment]
Hi @chgl ,
for some reason, the WEBAPI_URL was not getting populated in the atlas container.  I just pulled the scripts again and it is now showing up in the container without adding the extraENV value.

Also, what could have happened in my case is the TLS secret is something like "/Common/Bluemix"..  the values.yml spec did not allow me to put the /Common/Bluemix as the value as it failed regex validation.  Hence, I had to delete the ingress and recreate it manually.
